src/common/libflux/future.c

```diff
@@ -481,11 +483,21 @@ int flux_future_wait_for (flux_future_t *f, double timeout)
         f->now->running = false;
     }
     if (!future_is_ready (f)) {
-        /* Typically this error should be "impossible", but can occur
-         * if the future does not get initialized properly to use the
-         * now reactor.
+        /* Typically this error path should be "impossible", because
+         * the future should have been fulfilled or an error occurred
+         * in flux_reactor_run() above.
+         *
+         * However, if the future does not get initialized properly
+         * (i.e. via flux_future_create()'s init callback) the reactor
+         * may not have any watchers on it and return immediately,
+         * thus the future is still not ready.
+         *
+         * This is also why an "infinite timer watcher" is not created
+         * for the reactor in now_context_set_timeout().  In the event
+         * the flux future does not get initialized properly, a
+         * "infinite timer watcher" would lead to a hang / deadlock.
```

Comment:
The comment about an "infinite timer watcher" might be confusing here. At least I'm not sure what prompted this part of the comment and it would make me scratch my head. I would consider just dropping this section or replacing with something like "If the reactor did block without any watchers registered in a "now" context, then this would be deadlock. Therefore we return EDEADLOCK here to warn the user".